src/algorithms/tracking/AmbiguitySolver.cc

```diff
@@ -0,0 +1,266 @@
+          // FIXME why not pars.getCovariance()(i,j) = covariance(a,b) / x / y;
+          cov(i, j) = covariance(a, b) / x / y;
+        }
+      }
+      pars.setCovariance(cov);
+
+      trajectory.addToTrackParameters(pars);
+
+      // Fill tracks
+      auto track = tracks->create();
+      track.setType( // Flag that defines the type of track
+          pars.getType());
+      track.setPosition( // Track 3-position at the vertex
+          edm4hep::Vector3f());
+      track.setMomentum( // Track 3-momentum at the vertex [GeV]
+          edm4hep::Vector3f());
+      track.setPositionMomentumCovariance( // Covariance matrix in basis [x,y,z,px,py,pz]
+          edm4eic::Cov6f());
+      track.setTime( // Track time at the vertex [ns]
+          static_cast<float>(parameter[Acts::eBoundTime]));
+      track.setTimeError( // Error on the track vertex time
+          sqrt(static_cast<float>(covariance(Acts::eBoundTime, Acts::eBoundTime))));
+      track.setCharge( // Particle charge
+          std::copysign(1., parameter[Acts::eBoundQOverP]));
+      track.setChi2(trajectoryState.chi2Sum); // Total chi2
+      track.setNdf(trajectoryState.NDF);      // Number of degrees of freedom
+      track.setPdg(                           // PDG particle ID hypothesis
+          boundParam.particleHypothesis().absolutePdg());
+      track.setTrajectory(trajectory); // Trajectory of this track
+
+      mj.visitBackwards(trackTip, [&](const auto& state) {
+        auto geoID     = state.referenceSurface().geometryId().value();
+        auto typeFlags = state.typeFlags();
+
+        // find the associated hit (2D measurement) with state sourcelink index
+        // fix me: calibrated or not?
+        if (state.hasUncalibratedSourceLink()) {
+
+          std::size_t srclink_index = state.getUncalibratedSourceLink()
+                                          .template get<ActsExamples::IndexSourceLink>()
+                                          .index();
+
+          // no hit on this state/surface, skip
+          if (typeFlags.test(Acts::TrackStateFlag::HoleFlag)) {
+            m_log->debug("No hit found on geo id={}", geoID);
+
+          } else {
+            auto meas2D = meas2Ds[srclink_index];
+            if (typeFlags.test(Acts::TrackStateFlag::MeasurementFlag)) {
+              track.addToMeasurements(meas2D);
+              trajectory.addToMeasurements_deprecated(meas2D);
+              m_log->debug("Measurement on geo id={}, index={}, loc={},{}", geoID, srclink_index,
+                           meas2D.getLoc().a, meas2D.getLoc().b);
+
+            } else if (typeFlags.test(Acts::TrackStateFlag::OutlierFlag)) {
+              trajectory.addToOutliers_deprecated(meas2D);
+              m_log->debug("Outlier on geo id={}, index={}, loc={},{}", geoID, srclink_index,
+                           meas2D.getLoc().a, meas2D.getLoc().b);
+            }
+          }
+        }
+      });
```

Comment:
This code would need to be de-duplicated with the CKFTracking.cc.

Reply:
I suggest that we approve and merge #1470, and then refactor this code in a similar way to remove this chunk of code use another ActsToTracks_factory instances to perform the conversion.

Reply:
#1470 is merged. But by default it takes the first tracktip (as it was in tracking.cc), while in ambiguity solver it goes with a list of tip indices. @minjungkim12 @bschmookler we can add an optional input for ActsToTracks to take the tip index. But is it appropriate at all to take the first tip by default when no ambiguity solver is called?

Reply:
If we were to propagate list of tip indices it would have to be propagated to every consumer of `CentralCKFActsTrajectories`, which is not feasible. Would it be possible to remove/reorder tips to facilitate the expected behavior downstream?

Reply:
This is resolved by the update I made yesterday. The last comment is addressed in the conversation below.